### PR TITLE
Legalize large-stride non-overlapping maxpool on U55 (#21284)

### PR DESCRIPTION
**Pull request by apullin**
Summary:

Teach the Arm partitioner to recognize exact U55 large-stride non-overlapping max-pool operations as safely decomposable, without mutating the graph during partitioning. Backend preprocessing then canonicalizes value-only max_pool2d_with_indices and rewrites eligible pools into supported stride-1 reductions. This preserves the partitioner contract, keeps unrelated TOSA/VGF/U85 paths unchanged, and also covers exported MaxPool1d.

Differential Revision: D113087299

### CODE REVIEW

File: backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# Copyright 2026 Arm Limited and/or its affiliates.
```
Comment (rascani) [inline backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py]: Use Meta copyright. Arm will add theirs if/when they modify the file.

**Review by rascani — COMMENTED** [review]
@claude review this PR

**Comment by digantdesai** [comment]
@claude review this pr, since the previous attempt didn't work, let's try again

File: backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        and (dilation_h, dilation_w) == (1, 1)
+        and not ceil_mode
+        and 1 <= kernel_h <= 256
+        and 1 <= kernel_w <= 256
+        and height >= kernel_h
+        and width >= kernel_w
+    )
+
+
+class DecomposeLargeStrideMaxPool2dPass(ArmOpTargetedPass):
```
Comment (digantdesai) [inline backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py]: how does this work for non-U55? Mainly U85? Does it handle large strides natively?

File: backends/arm/_passes/arm_pass_manager.py [hunk 3]
```diff
@@ -69,6 +69,7 @@
     DecomposeIndexSelectToGatherPass,
     DecomposeIndexTensorToGatherPass,
     DecomposeIntPowPass,
+    DecomposeLargeStrideMaxPool2dPass,
```
Comment (digantdesai) [inline backends/arm/_passes/arm_pass_manager.py]:
```suggestion
    DecomposeLargeStrideMaxPool2dForU55Pass,
```

File: backends/arm/test/passes/test_decompose_large_stride_maxpool2d_pass.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+from types import SimpleNamespace
+from typing import Tuple
+from unittest.mock import patch
+
+import torch
+from executorch.backends.arm._passes.decompose_large_stride_maxpool2d_pass import (
+    DecomposeLargeStrideMaxPool2dPass,
+)
+from executorch.backends.arm._passes.remove_getitem_pass import RemoveGetItemPass
+from executorch.backends.arm.test.tester.test_pipeline import PassPipeline
```
Comment (digantdesai) [inline backends/arm/test/passes/test_decompose_large_stride_maxpool2d_pass.py]: should we also add EthosU55PipelineINT?

File: backends/arm/tosa/partitioner.py [hunk 5]
```diff
@@ -53,6 +56,36 @@
...
+        self.tosa_spec = tosa_spec
+
+    def is_node_supported(
+        self,
+        submodules: Mapping[str, torch.nn.Module],
+        node: torch.fx.Node,
+    ) -> bool:
+        """Return True when backend preprocessing can legalize the max pool."""
+        del submodules
+        if not self.tosa_spec.is_U55_subset or node.target not in {
```
Comment (digantdesai) [inline backends/arm/tosa/partitioner.py:81]: what about u85?

**Review by digantdesai — COMMENTED** [review]
Left some comments, mainly about broader implications of this decomp for u55. And also address Claude comments before landing.

File: backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py (hunk 2 above)
Comment (apullin) [reply backends/arm/_passes/decompose_large_stride_maxpool2d_pass.py]: This is U55-only. Non-U55 profiles should go through the standard TOSA/Vela maxpool path. The new pass and partitioner positive check are gated by `is_U55_subset`.
Renamed the pass/check to include `ForU55` and added comments to make that explicit.

File: backends/arm/test/passes/test_decompose_large_stride_maxpool2d_pass.py (hunk 4 above)
Comment (apullin) [reply backends/arm/test/passes/test_decompose_large_stride_maxpool2d_pass.py]: added a lightweight `EthosU55PipelineINT` coverage case with `run_on_fvp=False` to exercise the U55 INT lowering path without requiring FVP

File: backends/arm/tosa/partitioner.py (hunk 5 above)
Comment (apullin) [reply backends/arm/tosa/partitioner.py:81]: U85 should use the standard TOSA/Vela maxpool path. This workaround is only for the U55 stride > 3 pooling limitation, as that is a hardware limitation and U85 does not have that limit (as far as I understand it).

Renamed the pass/check to include ForU55 and kept the partitioner positive check gated on is_U55_subset, so U85 is unchanged.

**Comment by apullin** [comment]
@digantdesai

Addressed:

- Added U55 dim/kernel checks in `can_decompose_large_stride_maxpool2d`, including validation of the generated intermediate reduction shapes before partitioning accepts the node
- Renamed the pass/check to `*ForU55*` and documented that non-U55 profiles, including U85, use the standard TOSA/Vela maxpool path
- Documented the `max_pool2d_with_indices` contract: accepted because `RemoveGetItemPass` runs earlier and canonicalizes value-only users to `max_pool2d.default`; indices users are rejected by that pass, matching existing behavior
- Removed redundant SymInt re-check/recomputation in the pass body by centralizing validation/unpacking in the predicate/helper
- Kept intermediate qparams stripped intentionally; final output metadata carries original qparams. Added U55 INT pipeline coverage as a sanity check
- Expanded tests with numerical output comparison, non-U55 no-op coverage, `stride=None`, U55 limit rejection cases, and a lightweight `EthosU55PipelineINT(..., run_on_fvp=False)` case

**Comment by apullin** [comment]
OK, I believe lint is finally satisfied, too.

**Review by digantdesai — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.
